### ISSUE TITLE
Option to only download a portion of a project's data: Is there one? Should there be one?

### ISSUE THREAD

**Issue by jsayers-bsc** [MotusDev/Motus-TO-DO#767]
As I understand, downloading Project X data with the R package will result in all of that project's tag data being downloaded. Only after a local .motus file exists can data be subsetted by year, species, etc.

This works for most, but some projects generate a lot of data, for instance [nearly 52 million records in one year](https://github.com/MotusDev/Motus-TO-DO/issues/721). Members of this project, as well as myself, were having issues simply executing a filter query to reduce the size of the working dataset. I made a few mistakes that were contributing to my problems, like selecting the entire width of the alltags table instead just the desired columns, but I think it's still likely to be a challenge.

They actually just emailed about this as they are concerned they will likely add yet another 50 million records next year, and the next, etc. The question was whether they should just create a new project now so they can download this year's data and then join it in R to their filtered down data from 2019. In this case, the answer seems to be that yes, that's probably the best way for them to be able to work with their data as the size of the dataset grows.

So I have two questions:

- Am I missing something, and there is a way of downloading only a subset of a project's data?
- If not, _should_ we have a method of downloading only a portion of a project or receivers's data so that projects with 10s or 100s of millions of records, can start from a smaller local database?

**Comment by denislepage** [comment]
Correct, it’s currently all or nothing.

Adding some functionalities to filter would potentially be useful, yes, but not necessarily simple. The simplest way would be to limit the filtering on fields available in batches (receiver ID or batch start/end), or perhaps runs (tagId, run start/end).  I’m not saying no, but we’d carefully need to understand the implications, and which types of filtering people would want vs. what is reasonable for us to support. When we query by project, I know we first need to generate a list of batch ID’s, presumably by querying the run table. It may not be TOO complicated to add either/both the ability to restrict the data to a specific list of tags or a date range that limits the runs. We’d need to test both/either approaches for speed and complexity.

I know this is a bit beyond the current capacity of most users to go into details on this, but how you build your queries in your local motus file can make a very big difference in the response time, sometimes by several orders of magnitude. One of us (perhaps Adam?) had started looking at replacing the alltags view by dplyr syntax, which would allow people to better customize that they need. The alltags view is really just there for convenience, but if you are relying on this to only get a subset of fields, you may add a lot of overhead from joining tables that you may end up not using. Doing your own custom join of tables would be much better, but would need to be better documented for users. The GPS table join was definitely the biggest issue, but not the only one. I’m presuming that you were using the view without the GPS join?

If you want to share your R statements for filtering (and which project you were working with), I can better understand.

**Comment by jsayers-bsc** [comment]
This is still Project 279, same as the linked post regarding [working with large datasets](https://github.com/MotusDev/Motus-TO-DO/issues/721), and I haven't dug into their project again following their most recent question.

As you mention, the alltags view is part of the issue here, ironically because of its convenience. It's easy to think of it as just another table so most people aren't aware that behind the scenes it's actually a complex set of table joins all happening in memory. It's very convenient for the typical dataset but as they grow in size it becomes very unwieldy (even without the GPS). 

Then the logical step of only selecting the desired fields and filtering a subset of date prior to collapsing into a dataframe isn't helping either. First because those queries all just get queued up as one database call at the last moment (unless intermediately collecting to a dataframe, which won't always be possible for very large datasets) and second because all of this has to happen _after_ all the table joins occur to create the view in the first place — all in memory! 

It sounds like doing your own table joins, using only the fields needed, would probably be more ideal in these cases. 
But because alltags is so convenient, many people probably don't dig much deeper into the underlying tables that make up the alltags view (for that matter, I'm not sure the R Book actually goes into much depth about them either, expect where they show up in examples). 

Ok that helps clarify this quite a bit, as well as my previous general question about working with very large datasets.

So the question is, if eschewing the alltags view and working directly with just the tables and fields required, do any of you have some sense for how large a Motus dataset has been or could be worked with? Project 279 expects 150 million records and a .motus file 18gb in size in another 2 years. How feasible will that be?

**Comment by adamdsmith** [comment]
Not entirely sure it's helpful, but if it comes to building your own `alltags` or something other custom joins, [I found dramatic speed improvements using the `dm` package](https://github.com/MotusDev/Motus-TO-DO/issues/645#issuecomment-570571301). I never perfectly replicated `alltags`, but I got close. I've not had time to revisit to see if `alltags` could be replicated perfectly. But if you're targeting some other custom joins that doesn't replicate `alltags`, the `dm` approach may be considerably faster.
